File: app/src/main/java/com/scullyapps/spoonsbottleup/activities/FridgeManagementActivity.java
package com.scullyapps.spoonsbottleup.activities;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Bundle;
import android.util.Log;

import com.scullyapps.spoonsbottleup.R;
import com.scullyapps.spoonsbottleup.data.BottleDatabase;
import com.scullyapps.spoonsbottleup.models.Bottle;
import com.scullyapps.spoonsbottleup.ui.Fridge;
import com.scullyapps.spoonsbottleup.ui.fridgeman.ItemTouchCallback;
import com.scullyapps.spoonsbottleup.ui.fridgeman.RecyclerListAdapter;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class FridgeManagementActivity extends AppCompatActivity {

    @BindView(R.id.fridgeman_recycler)
    RecyclerView list;

    ArrayList<Bottle> bottles;

    List<Bottle> toRemove = new ArrayList<>();

    String fridgeName;

    private ItemTouchHelper itemTouchHelper;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_fridge_management);

        ButterKnife.bind(this);

        Bundle b = getIntent().getExtras();

        Fridge fridge;


        Toolbar toolbar = findViewById(R.id.toolbar_fridgeman);

        if(b != null) {
            fridgeName = b.getString("name", "NONE");

            fridge = new Fridge(this, fridgeName);

            fridge.setBottles(BottleDatabase.INSTANCE.getBottlesByFridge(b.getString("name")));

            bottles = BottleDatabase.INSTANCE.getBottlesByFridge(b.getString("name"));

            toolbar.setTitle("Editing " + fridgeName);
        }

        RecyclerListAdapter adapter = new RecyclerListAdapter(bottles);

        ItemTouchHelper.Callback callback = new ItemTouchCallback(adapter);

        itemTouchHelper = new ItemTouchHelper(callback);

        list = (RecyclerView) findViewById(R.id.fridgeman_recycler);

        list.setHasFixedSize(true);
        list.setAdapter(adapter);
        list.setLayoutManager(new LinearLayoutManager(this));

        itemTouchHelper.attachToRecyclerView(list);

        adapter.setIth(itemTouchHelper);




        for(int i = 0; i < bottles.size(); i++) {
            Bottle btl = bottles.get(i);
        }
    }

    @Override
    protected void onDestroy() {
        for(Bottle bottle : toRemove) {
            Log.d("DEBUG", "onDestroy: removing " + bottle.getName());
        }

        super.onDestroy();
    }

    @Override
    public void onBackPressed() {

        RecyclerListAdapter adapter = (RecyclerListAdapter) list.getAdapter();


        // if no changes made, then we can safely exit
        if(!adapter.modified) {
            super.onBackPressed();
            return;
        }


        AlertDialog confirm = new AlertDialog.Builder(this)
                .setMessage("Do you wish to save changes?")
                .setPositiveButton("Save changes", (dialog, which) -> {
                    List<Bottle> toChange = adapter.getItems();

                    if(toChange == null)
                        super.onBackPressed();

                    for(int i = 0; i < toChange.size(); i++) {

                        Bottle current = toChange.get(i);
                        BottleDatabase.INSTANCE.updateListOrder(i, current.getId());
                        BottleDatabase.INSTANCE.updateListOrder(99, 5);
                    }

                    super.onBackPressed();
                    dialog.cancel();
                })
                .setNegativeButton("Discard", (dialog, which) -> {
                    toRemove.clear();
                    super.onBackPressed();
                    dialog.cancel();
                })
                .create();

        confirm.show();
    }
}

File: app/src/main/java/com/scullyapps/spoonsbottleup/activities/FridgeManagementActivity.kt
package com.scullyapps.spoonsbottleup.activities

import android.os.Bundle
import android.util.Log
import android.view.View
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import com.scullyapps.spoonsbottleup.R
import com.scullyapps.spoonsbottleup.adapters.BottleRecyclerAdapter
import com.scullyapps.spoonsbottleup.data.BottleDatabase
import com.scullyapps.spoonsbottleup.models.Bottle
import com.scullyapps.spoonsbottleup.models.Fridge
import com.scullyapps.spoonsbottleup.ui.dialogs.EditBottleDialog
import com.scullyapps.spoonsbottleup.ui.fridgeman.ItemTouchCallback
import kotlinx.android.synthetic.main.activity_fridge_management.*

class FridgeManagementActivity : AppCompatActivity() {

    var bottles: List<Bottle> = emptyList()
    var toRemove: MutableList<Bottle> = ArrayList()

    private val TAG = "FridgeManagementActivity"

    lateinit var database: BottleDatabase

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_fridge_management)

        // initialize database
        database = BottleDatabase.getInstance(this)

        val bundle = intent.extras
        val fridge: Fridge
        val toolbar = findViewById<Toolbar>(R.id.toolbar_fridgeman)

        toolbar?.inflateMenu(R.menu.menu_settings_bottles)
        toolbar?.setOnMenuItemClickListener { menuItem ->

            when(menuItem.itemId) {
                R.id.action_add_bottle -> {
                    var bottleToAdd = Bottle(
                        0,
                        "New Bottle",
                        "",
                        "",
                        2,
                        -1,
                        "Cupboard",
                        0,
                        0,
                        isDrink = false,
                        isMisc = false,
                        isDraught = false,
                        isSpirit = false,
                        isCanOrBottle = true
                    )

                    val dialog = EditBottleDialog(this, bottleToAdd)
                    dialog.setPositiveButton("Create", null)
                    dialog.onSubmitted = { bottle ->
                        database.bottleRoomDao.insert(bottle)
                    }

                    dialog.show()
                }
            }
            false
        }



        if (bundle != null) {
            // retrieve our fridge
            val fridgeName = bundle.getString("name", "NONE")

            fridge = database.fridgeRoomDao.query(fridgeName)
            bottles = database.bottleRoomDao.queryBottlesByFridge(fridge.name)

            toolbar.title = "Editing $fridgeName"

            val adapter = BottleRecyclerAdapter(bottles as ArrayList<Bottle>, fridge.name)
            val callback: ItemTouchHelper.Callback = ItemTouchCallback(adapter)
            val itemTouchHelper = ItemTouchHelper(callback)

//            fridgeman_recycler.setHasFixedSize(true)
            fridgeman_recycler.adapter = adapter
            fridgeman_recycler.layoutManager = LinearLayoutManager(this)

            itemTouchHelper.attachToRecyclerView(fridgeman_recycler)
            adapter.setTouchHelper(itemTouchHelper)

            fridgeman_nodata_text.text = getString(R.string.fridge_management_no_data, fridge.name)

            if(bottles.isEmpty()) {
                fridgeman_nodata.visibility = View.VISIBLE
            } else {
                fridgeman_nodata.visibility = View.GONE
            }
        }


    }



    // since the adapter will map any changes to the order,
    // we can just write the current order of items back to the db
    private fun saveOrder() {
        val adapter = fridgeman_recycler.adapter as BottleRecyclerAdapter

        for(i in 0 until adapter.items.size) {
            val bottle = adapter.items[i]

            Log.d(TAG, "Updating (${bottle.name}) ${bottle.ID} to LO: $i")

            bottle.listOrder = i

            run {
                database.bottleRoomDao.update(bottle)
            }
        }
    }

    override fun onBackPressed() {
        val adapter = fridgeman_recycler.adapter as BottleRecyclerAdapter

        // if no changes made, then we can safely exit
        if (!adapter.modified) {
            super.onBackPressed()
            return
        }

        val confirm = AlertDialog.Builder(this)
                .setMessage("Do you wish to save changes?")
                .setPositiveButton("Save changes") { dialog, _ ->
                    saveOrder()
                    super.onBackPressed()
                    dialog.cancel()
                }
                .setNegativeButton("Discard") { dialog, _ ->
                    toRemove.clear()
                    super.onBackPressed()
                    dialog.cancel()
                }
                .create()
        confirm.show()
    }
}